pybamm/solvers/base_solver.py

```diff
@@ -1425,6 +1429,52 @@ def _set_up_model_inputs(model, inputs):
 
         return ordered_inputs
 
+    def _compare_events_to_output_variables(self, model, output_variables):
+        """
+        Compares the model events and output variables requested by the user to check
+        that all events can be calculated using the output variables which will be
+        returned by the solver.
+        If the event doesn't require access to the state vector ('y[...]') then it
+        doesn't need a corresponding output variable.
+
+        Parameters
+        ----------
+        model : :class:`pybamm.BaseModel`
+            The model object
+        output_variables : list[str]
+            List of variables that will be returned by the solver
+        """
+
+        output_vars = {var: model.variables[var] for var in output_variables}
+
+        for event in model.events:
+            event_children = event.expression.children
+
+            if all(
+                [
+                    "y[" not in event_children[i].__str__()
+                    for i in range(len(event_children))
+                ]
+            ):
+                # event can be calculated without state vector, variables not needed
+                continue
+
+            # check for variables that match the event children in output_variables
+            matching_var = {k: v for k, v in output_vars.items() if v in event_children}
+
+            # search for variables that can be used to calculate the event
+            if len(matching_var) == 0:
+                required_vars = []
+                for k, v in model.variables.items():
+                    if v in event_children:
+                        required_vars.append(k)
+                joined_vars = "\n".join(required_vars)
+                raise ValueError(
```

Comment:
Would it be better to raise a [warning](https://docs.python.org/3/library/warnings.html) here rather than an Error, as this lets the user decide whether identification of the specific termination event is worth the expense of adding multiple additional variables (and their sensitivities) to the solution.

Reply:
Raising a warning and filling in the termination event with an empty string/"unknown" or similar would work for single simulations, but when used with an experiment the termination event is required to check when to move through experiment steps. This would require adding in different behaviour for using the solver with/without an experiment, perhaps auto-adding appropriate output variables when used with experiments. Currently sensitivities can't be used with experiments so there's minimal change to the computational expense.

An alternative implementation to consider in the future might be to change the IDAKLU solver output to include the state vector for the final timestep in the return value. This shouldn't impact the memory usage, and would allow calculation of any termination events using the pre-existing methods without the additional overhead of storing and calculating unwanted output variables/sensitivities during the solve.

Reply:
Sounds good!

Reply:
> An alternative implementation to consider in the future might be to change the IDAKLU solver output to include the state vector for the final timestep in the return value. This shouldn't impact the memory usage, and would allow calculation of any termination events using the pre-existing methods without the additional overhead of storing and calculating unwanted output variables/sensitivities during the solve.

This is a great solution for all solvers since the final state vector is required to initialize the next step in an experiment.